File: python/cofiswarm_observer/servicecomponent.py
"""Canonical async NATS service component for cofiswarm Python components — the Python port of
cofiswarm-observer-sdk/pkg/servicecomponent. Announces presence on the observer bus, serves
capability subjects behind the schema-major gate with loud-error replies, re-announces on
hello, and says goodbye on shutdown.

A route handler is an async callable taking the decoded request dict and returning a reply
dict; schema_version and ok are filled in automatically if the handler omits them.
"""
from __future__ import annotations

import json
import logging
from typing import Awaitable, Callable

import nats

from . import contract

logger = logging.getLogger("cofiswarm.observer")

Handler = Callable[[dict], Awaitable[dict]]


class ServiceComponent:
    """A generic bus capability process: announce + serve subjects + hello/goodbye."""

    def __init__(self, nc, name: str, kind: str, routes: dict[str, Handler]):
        self._nc = nc
        self._name = name
        self._kind = kind
        self._cid = kind  # component_id defaults to kind, mirroring the Go component
        self._routes = routes
        self._primary = sorted(routes)[0] if routes else ""

    @classmethod
    async def connect(cls, url: str, name: str):
        """Open a NATS connection with infinite reconnect (broker-bounce resilience)."""
        return await nats.connect(url, name=name, max_reconnect_attempts=-1)

    async def start(self) -> None:
        """Subscribe the routes plus hello (for re-announce) and publish the initial announce."""
        for subject, handler in self._routes.items():
            await self._nc.subscribe(subject, cb=self._make_cb(subject, handler))
        await self._nc.subscribe(contract.SUBJ_HELLO, cb=self._on_hello)
        await self._announce()
        logger.info(
            "%s (%s) serving %d subjects (id=%s)",
            self._name, self._kind, len(self._routes), self._cid,
        )

    def _make_cb(self, subject: str, handler: Handler):
        async def cb(msg) -> None:
            try:
                raw = json.loads(msg.data)
            except (ValueError, TypeError) as exc:
                logger.error("%s dropped malformed message on %s: %s", self._name, subject, exc)
                await self._reply_err(msg, "invalid json")
                return
            if not contract.major_supported(raw):
                logger.error("%s rejected %s: unsupported schema_version", self._name, subject)
                await self._reply_err(msg, "unsupported schema_version")
                return
            try:
                reply = await handler(raw)
            except Exception as exc:  # loud: a handler failure replies an error, never silent
                logger.error("%s handler %s failed: %s", self._name, subject, exc)
                await self._reply_err(msg, str(exc))
                return
            reply.setdefault("schema_version", contract.SCHEMA_VERSION)
            reply.setdefault("ok", True)
            await self._respond(msg, reply)

        return cb

    async def _on_hello(self, msg) -> None:
        logger.info("%s re-announcing on hello", self._name)
        await self._announce()

    async def _reply_err(self, msg, message: str) -> None:
        await self._respond(msg, {
            "schema_version": contract.SCHEMA_VERSION, "ok": False, "error": message,
        })

    async def _respond(self, msg, payload: dict) -> None:
        if not getattr(msg, "reply", ""):
            return
        try:
            await self._nc.publish(msg.reply, json.dumps(payload).encode())
        except Exception as exc:
            logger.error("%s failed to respond on %s: %s", self._name, msg.subject, exc)

    async def _announce(self) -> None:
        announce = {
            "schema_version": contract.SCHEMA_VERSION,
            "component_id": self._cid,
            "kind": self._kind,
            "info": {"name": self._name, "engine": self._kind, "tags": ["resource"]},
            "infer_subject": self._primary,
        }
        try:
            await self._nc.publish(contract.SUBJ_ANNOUNCE, json.dumps(announce).encode())
        except Exception as exc:
            logger.error("%s announce failed: %s", self._name, exc)

    async def shutdown(self) -> None:
        """Publish a graceful goodbye so presence flips offline quietly."""
        goodbye = {
            "schema_version": contract.SCHEMA_VERSION, "component_id": self._cid, "reason": "shutdown",
        }
        try:
            await self._nc.publish(contract.SUBJ_GOODBYE, json.dumps(goodbye).encode())
            await self._nc.flush()
        except Exception as exc:
            logger.error("%s goodbye failed: %s", self._name, exc)

File: pkg/servicecomponent/servicecomponent.go
// Package servicecomponent is the canonical NATS "service component" for cofiswarm: it
// announces presence on the observer bus, serves a set of capability subjects behind a
// schema-major gate with loud-error replies, re-announces on hello, and says goodbye on
// shutdown. It replaces the byte-identical per-repo internal/bus copies (launcher,
// slot-manager, kvpool) with one shared source, so the wire schema can no longer drift.
package servicecomponent

import (
	"encoding/json"
	"fmt"
	"log"
	"sort"

	"github.com/nats-io/nats.go"

	"github.com/keepdevops/cofiswarm-observer-sdk/pkg/contract"
)

// Re-exported so callers that only build replies/subjects need just this package.
const (
	SchemaVersion = contract.SchemaVersion
	Prefix        = contract.Prefix
)

type modelInfo struct {
	Name   string   `json:"name"`
	Engine string   `json:"engine"`
	Tags   []string `json:"tags,omitempty"`
}

type announceMsg struct {
	SchemaVersion string    `json:"schema_version"`
	ComponentID   string    `json:"component_id"`
	Kind          string    `json:"kind"`
	Info          modelInfo `json:"info"`
	InferSubject  string    `json:"infer_subject"`
}

type goodbyeMsg struct {
	SchemaVersion string `json:"schema_version"`
	ComponentID   string `json:"component_id"`
	Reason        string `json:"reason"`
}

// Handler decodes a request and returns a reply struct; a non-nil error is replied loud.
type Handler func(data []byte) (any, error)

// Component is a generic bus capability process (the Go ServiceComponent).
type Component struct {
	nc      *nats.Conn
	name    string
	kind    string
	cid     string
	routes  map[string]Handler
	primary string
}

// Connect opens a NATS connection with infinite reconnect (broker-bounce resilience).
func Connect(url, name string) (*nats.Conn, error) {
	return nats.Connect(url, nats.Name(name), nats.MaxReconnects(-1))
}

// New builds a component serving routes. The lexically-first subject is the announced primary.
func New(nc *nats.Conn, name, kind string, routes map[string]Handler) *Component {
	keys := make([]string, 0, len(routes))
	for s := range routes {
		keys = append(keys, s)
	}
	sort.Strings(keys) // deterministic primary subject for the announce
	primary := ""
	if len(keys) > 0 {
		primary = keys[0]
	}
	return &Component{nc: nc, name: name, kind: kind, cid: kind, routes: routes, primary: primary}
}

// Start subscribes the routes plus hello (for re-announce) and publishes the initial announce.
func (c *Component) Start() error {
	for subj, h := range c.routes {
		if _, err := c.nc.Subscribe(subj, c.makeCB(subj, h)); err != nil {
			return fmt.Errorf("subscribe %s: %w", subj, err)
		}
	}
	if _, err := c.nc.Subscribe(contract.SubjHello, func(*nats.Msg) { c.announce() }); err != nil {
		return fmt.Errorf("subscribe hello: %w", err)
	}
	c.announce()
	log.Printf("%s (%s) serving %d subjects (id=%s)", c.name, c.kind, len(c.routes), c.cid)
	return nil
}

func (c *Component) makeCB(subj string, h Handler) nats.MsgHandler {
	return func(m *nats.Msg) {
		var raw map[string]json.RawMessage
		if err := json.Unmarshal(m.Data, &raw); err != nil {
			log.Printf("%s dropped malformed message on %s", c.name, subj)
			c.replyErr(m, "invalid json")
			return
		}
		if !contract.MajorSupported(raw) {
			log.Printf("%s rejected %s: unsupported schema_version", c.name, subj)
			c.replyErr(m, "unsupported schema_version")
			return
		}
		reply, err := h(m.Data)
		if err != nil {
			log.Printf("%s handler %s failed: %v", c.name, subj, err)
			c.replyErr(m, err.Error())
			return
		}
		b, err := json.Marshal(reply)
		if err != nil {
			log.Printf("%s marshal reply failed on %s: %v", c.name, subj, err)
			c.replyErr(m, "marshal reply")
			return
		}
		_ = m.Respond(b)
	}
}

func (c *Component) replyErr(m *nats.Msg, msg string) {
	b, _ := json.Marshal(map[string]any{
		"schema_version": contract.SchemaVersion, "ok": false, "error": msg,
	})
	_ = m.Respond(b)
}

func (c *Component) announce() {
	a := announceMsg{
		SchemaVersion: contract.SchemaVersion, ComponentID: c.cid, Kind: c.kind,
		Info:         modelInfo{Name: c.name, Engine: c.kind, Tags: []string{"resource"}},
		InferSubject: c.primary,
	}
	if b, err := json.Marshal(a); err == nil {
		_ = c.nc.Publish(contract.SubjAnnounce, b)
	}
}

// Shutdown publishes a graceful goodbye so presence flips offline quietly.
func (c *Component) Shutdown() {
	b, _ := json.Marshal(goodbyeMsg{SchemaVersion: contract.SchemaVersion, ComponentID: c.cid, Reason: "shutdown"})
	_ = c.nc.Publish(contract.SubjGoodbye, b)
}
